src/switchroot/ostree-prepare-root.c

```diff
@@ -274,53 +280,76 @@ validate_signature (GBytes *data, GVariant *signatures, const guchar *pubkey, si
 typedef struct
 {
   OtTristate enabled;
+  gboolean is_signed;
   char *signature_pubkey;
-  char *expected_digest;
+  GList *pubkeys;
 } ComposefsConfig;
 
 static void
 free_composefs_config (ComposefsConfig *config)
 {
   free (config->signature_pubkey);
-  free (config->expected_digest);
   free (config);
 }
 
 G_DEFINE_AUTOPTR_CLEANUP_FUNC (ComposefsConfig, free_composefs_config)
 
 static ComposefsConfig *
-load_composefs_config (GError **error)
+load_composefs_config (GKeyFile *config, GError **error)
 {
   GLNX_AUTO_PREFIX_ERROR ("Loading composefs config", error);
+
   g_autoptr (ComposefsConfig) ret = g_new0 (ComposefsConfig, 1);
-  ret->enabled = OT_TRISTATE_MAYBE;
 
-  // TODO: Drop this kernel argument in favor of just the config file in the initramfs
-  autofree char *ot_composefs = read_proc_cmdline_key (OT_COMPOSEFS_KARG);
-  if (ot_composefs)
+  g_autofree char *enabled = g_key_file_get_value (config, COMPOSEFS_KEY, ENABLED_KEY, NULL);
+  if (g_strcmp0 (enabled, "signed") == 0)
     {
-      if (strcmp (ot_composefs, "off") == 0)
-        ret->enabled = OT_TRISTATE_NO;
-      else if (strcmp (ot_composefs, "maybe") == 0)
-        ret->enabled = OT_TRISTATE_MAYBE;
-      else if (strcmp (ot_composefs, "on") == 0)
-        ret->enabled = OT_TRISTATE_YES;
-      else if (g_str_has_prefix (ot_composefs, "signed="))
-        {
-          ret->enabled = OT_TRISTATE_YES;
-          ret->signature_pubkey = g_strdup (ot_composefs + strlen ("signed="));
-        }
-      else if (g_str_has_prefix (ot_composefs, "digest="))
+      ret->enabled = OT_TRISTATE_YES;
+      ret->is_signed = true;
+    }
+  else if (!ot_keyfile_get_tristate_with_default (config, COMPOSEFS_KEY, ENABLED_KEY,
+                                                  OT_TRISTATE_MAYBE, &ret->enabled, error))
+    return NULL;
+
+  if (!ot_keyfile_get_value_with_default (config, COMPOSEFS_KEY, KEYPATH_KEY, DEFAULT_KEYPATH,
+                                          &ret->signature_pubkey, error))
+    return NULL;
+
+  if (ret->is_signed)
+    {
+      g_autofree char *pubkeys = NULL;
+      gsize pubkeys_size;
+
+      /* Load keys */
+
+      if (!g_file_get_contents (ret->signature_pubkey, &pubkeys, &pubkeys_size, error))
+        return glnx_prefix_error_null (error, "Reading public key file '%s'",
+                                       ret->signature_pubkey);
+
+      /* Raw binary form if right size */
```

Comment:
Having two different file formats seems like it needs some motivation.  If we support this we should document it.  I'd vote we drop it and only support base64+newline separated format, which is the same thing supported by the main ostree+ed25519 format.

(There's a larger picture question here of whether we should actually also leverage openssl here and parse PEM for example too)

Reply:
Both file formats are supported by ostree_sign_ed25519_add_pk() though, based on the variant type (bytestring vs string).

Reply:
I guess for the ostree sign --keys-file we only ever use the base64 version though.
The current format type in the autosd images is raw, but we can easily change this.